**Be sure to save a copy of this notebook in your Google Drive and edit your saved copy.**

## Overview

In this PA, you will will be creating and manipulating basic objects to uncover a top secret message.

### Hint 1

As you work through this PA, you will encounter some code that does not work as you want it to.  Don't despair!  Errors (when Python is unable to run your code) and bugs (when the code doesn't do what you hoped) are a natural part of coding.  Even the best of the best deal with these problems regularly - learning to track down the issue is a skill that you can learn and practice.

Google is your friend!  If an error is confusing, copy it into Google and see what other people are saying.  If you don't know how to do something, search for it.

### Hint 2

Just because there is no error message doesn't mean everything went smoothly.  You should be regularly printing out your objects, to see they contain the information and format that you expect.

For example, if you have an object named `my_thing`, you might first want to simply look at the object by typing its name.

However, you might also want to make sure it is the right **object type** and/or the right **object structure**.


In [48]:
type(my_thing)
len(my_thing)

NameError: name 'my_thing' is not defined

## Part One: Setup

Each of the following code chunks will cause an error and/or do the desired task incorrectly.  Find the mistake, and correct it to complete the intended action.




1. Create lists containing the upper case letters, lower case letters, and some punctuation marks.


In [1]:
lower_case = ["a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s", "t", "u", "v", "w", "x", "y", "z"]

upper_case = ["A", "B", "C", "D", "E", "F", "G", "H" "I", "J", "K", "L", "M", "N", "O", "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z"]

punctuation = [".", ",", "!", "?", "'", "\"", "(", ")", " ", "-", ";", ":"]

2. Make one long list containing all the symbols.  Find the length of this list.


In [2]:
all_symbols = lower_case + upper_case + punctuation

len(all_symbols)



63

3. Repeat the previous steps, but instead of using *list* type objects, use **numpy arrays**.  Don't forget to **import** the numpy library!


In [3]:
import numpy as np

lower_case = np.array(["a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s", "t", "u", "v", "w", "x", "y", "z"])

upper_case = np.array(["A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M", "N", "O", "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z"])

punctuation = np.array([".", ",", "!", "?", "'", "\"", "(", ")", " ", "-", ";", ":"])

## Use np.concatenate to combine into long array
all_symbols = np.concatenate ([lower_case, upper_case, punctuation])

4. Make a **pandas data frame**, with one column called "Symbol" containing the `all_symbols` array.


In [19]:
import pandas as pd

symbols_df = pd.DataFrame(all_symbols)

symbols_df = symbols_df.rename(columns={0: "Symbol"})

print(symbols_df.columns)

Index(['Symbol'], dtype='object')


## Part Two: Decoding the secret message.


In [5]:
top_secret = pd.read_csv("https://www.dropbox.com/s/k72h1zewk4gtqep/PA_Secret_Code?dl=1", header=None)

top_secret = top_secret[0].to_numpy()
top_secret = top_secret + 14
top_secret = (top_secret * 18) - 257
top_secret = np.exp(top_secret)
top_secret = top_secret ** 2

print(top_secret)




[5.900e+01 5.000e+01 2.900e+01 9.000e+00 2.900e+01 7.300e+01 1.600e+01 3.400e+01 2.700e+01
 3.200e+01 3.400e+01 3.300e+01 3.300e+01 8.200e+01 9.100e+01 3.300e+01 4.800e+01 4.000e+01
 9.900e+01 7.900e+01 5.000e+01 4.900e+01 1.070e+02 6.700e+01 6.200e+01 6.100e+01 9.300e+01
 6.400e+01 8.300e+01 1.210e+02 1.010e+02 7.900e+01 8.800e+01 7.300e+01 8.900e+01 1.330e+02
 1.350e+02 3.797e+03 9.780e+02 1.610e+02 9.800e+01 3.805e+03 1.350e+02 7.130e+02 4.140e+02
 1.170e+02 3.815e+03 9.700e+01 2.940e+02 1.160e+02 3.823e+03 1.530e+02 1.870e+02 2.770e+02
 1.140e+02 2.560e+02 1.390e+02 3.837e+03 1.990e+02 3.160e+02 3.843e+03 1.645e+03 1.900e+02
 1.530e+02 3.851e+03 6.610e+02 1.350e+02 1.400e+02 1.630e+02 4.236e+03 3.863e+03 8.730e+02
 2.900e+02 2.920e+02 3.871e+03 3.210e+02 2.350e+02 3.250e+02 5.190e+02 7.850e+02 3.883e+03
 6.930e+02 1.910e+02 4.920e+02 1.950e+02 3.893e+03 1.695e+03 2.400e+02 2.030e+02 3.901e+03
 1.860e+02 4.090e+02 5.100e+02 4.130e+02 2.390e+02 4.170e+02 6.780e+02 2.210e+02 5.590e+02

In [6]:
nums_below_17 = np.sum(top_secret < 17)
print("Numbers below 17:", nums_below_17)

Numbers below 17: 352


In [7]:
top_secret_matrix = top_secret.reshape(-1, 5, order="F")
print(top_secret_matrix)

[[5.900e+01 1.440e+02 3.610e+02 9.000e+00 2.050e+01]
 [5.000e+01 1.000e+00 1.521e+03 7.500e+00 3.050e+01]
 [2.900e+01 5.290e+02 2.250e+02 1.050e+01 3.000e+00]
 [9.000e+00 3.610e+02 2.250e+02 3.500e+00 9.000e+00]
 [2.900e+01 3.721e+03 1.600e+01 4.000e+00 5.000e-01]
 [7.300e+01 1.521e+03 3.721e+03 3.050e+01 1.000e+00]
 [1.600e+01 6.400e+01 1.000e+00 5.000e-01 5.000e+00]
 [3.400e+01 1.000e+00 5.290e+02 7.000e+00 7.500e+00]
 [2.700e+01 1.521e+03 6.400e+01 2.000e+00 1.050e+01]
 [3.200e+01 3.721e+03 8.100e+01 3.050e+01 9.500e+00]
 [3.400e+01 9.000e+00 1.440e+02 1.950e+01 3.050e+01]
 [3.300e+01 1.000e+00 2.500e+01 4.000e+00 2.000e+00]
 [3.300e+01 1.521e+03 3.721e+03 9.000e+00 5.000e-01]
 [8.200e+01 9.000e+00 8.100e+01 7.500e+00 1.250e+01]
 [9.100e+01 6.400e+01 1.960e+02 1.050e+01 2.750e+01]
 [3.300e+01 3.025e+03 3.721e+03 3.500e+00 3.050e+01]
 [4.800e+01 3.721e+03 1.521e+03 4.000e+00 1.450e+01]
 [4.000e+01 7.840e+02 1.640e+02 3.050e+01 5.000e-01]
 [9.900e+01 2.500e+01 3.250e+02 3.050e+01 6.00

In [8]:
even_nums = np.arange(2, 383, 2)
print(even_nums)

[  2   4   6   8  10  12  14  16  18  20  22  24  26  28  30  32  34  36  38  40  42  44
  46  48  50  52  54  56  58  60  62  64  66  68  70  72  74  76  78  80  82  84  86  88
  90  92  94  96  98 100 102 104 106 108 110 112 114 116 118 120 122 124 126 128 130 132
 134 136 138 140 142 144 146 148 150 152 154 156 158 160 162 164 166 168 170 172 174 176
 178 180 182 184 186 188 190 192 194 196 198 200 202 204 206 208 210 212 214 216 218 220
 222 224 226 228 230 232 234 236 238 240 242 244 246 248 250 252 254 256 258 260 262 264
 266 268 270 272 274 276 278 280 282 284 286 288 290 292 294 296 298 300 302 304 306 308
 310 312 314 316 318 320 322 324 326 328 330 332 334 336 338 340 342 344 346 348 350 352
 354 356 358 360 362 364 366 368 370 372 374 376 378 380 382]


In [9]:
top_secret_matrix[:, 0] = top_secret_matrix[:, 0] - even_nums
print(top_secret_matrix[:, 0] )

[5.700e+01 4.600e+01 2.300e+01 1.000e+00 1.900e+01 6.100e+01 2.000e+00 1.800e+01 9.000e+00
 1.200e+01 1.200e+01 9.000e+00 7.000e+00 5.400e+01 6.100e+01 1.000e+00 1.400e+01 4.000e+00
 6.100e+01 3.900e+01 8.000e+00 5.000e+00 6.100e+01 1.900e+01 1.200e+01 9.000e+00 3.900e+01
 8.000e+00 2.500e+01 6.100e+01 3.900e+01 1.500e+01 2.200e+01 5.000e+00 1.900e+01 6.100e+01
 6.100e+01 3.721e+03 9.000e+02 8.100e+01 1.600e+01 3.721e+03 4.900e+01 6.250e+02 3.240e+02
 2.500e+01 3.721e+03 1.000e+00 1.960e+02 1.600e+01 3.721e+03 4.900e+01 8.100e+01 1.690e+02
 4.000e+00 1.440e+02 2.500e+01 3.721e+03 8.100e+01 1.960e+02 3.721e+03 1.521e+03 6.400e+01
 2.500e+01 3.721e+03 5.290e+02 1.000e+00 4.000e+00 2.500e+01 4.096e+03 3.721e+03 7.290e+02
 1.440e+02 1.440e+02 3.721e+03 1.690e+02 8.100e+01 1.690e+02 3.610e+02 6.250e+02 3.721e+03
 5.290e+02 2.500e+01 3.240e+02 2.500e+01 3.721e+03 1.521e+03 6.400e+01 2.500e+01 3.721e+03
 4.000e+00 2.250e+02 3.240e+02 2.250e+02 4.900e+01 2.250e+02 4.840e+02 2.500e+01 3.610e+02

In [10]:
top_secret_matrix[17:24,2] = top_secret_matrix[17:24,2] - 100
print (top_secret_matrix[17:24,2])

[  64.  225.  441.   49.   64. 1521. 2809.]


In [11]:
top_secret_matrix[:, 3:5] = top_secret_matrix[:, 3:5] * 2
print (top_secret_matrix[:, 3:5])

[[18. 41.]
 [15. 61.]
 [21.  6.]
 [ 7. 18.]
 [ 8.  1.]
 [61.  2.]
 [ 1. 10.]
 [14. 15.]
 [ 4. 21.]
 [61. 19.]
 [39. 61.]
 [ 8.  4.]
 [18.  1.]
 [15. 25.]
 [21. 55.]
 [ 7. 61.]
 [ 8. 29.]
 [61.  1.]
 [61. 12.]
 [61. 12.]
 [46. 15.]
 [ 8. 15.]
 [ 5.  8.]
 [61. 55.]
 [22. 61.]
 [15. 29.]
 [18.  1.]
 [16. 12.]
 [ 1. 12.]
 [12.  1.]
 [61. 25.]
 [ 2. 55.]
 [12. 58.]
 [ 1. 61.]
 [ 4. 61.]
 [ 5. 61.]
 [61. 34.]
 [23.  5.]
 [ 5. 61.]
 [14.  3.]
 [39.  8.]
 [61. 15.]
 [19. 18.]
 [14. 39.]
 [ 9. 12.]
 [ 3.  5.]
 [11.  4.]
 [ 5. 61.]
 [18.  9.]
 [62. 14.]
 [19. 61.]
 [14.  8.]
 [ 1.  9.]
 [ 3. 19.]
 [11. 61.]
 [55. 10.]
 [61. 15.]
 [34. 25.]
 [ 5. 53.]
 [61. 61.]
 [12. 57.]
 [ 5. 46.]
 [ 6. 23.]
 [39.  1.]
 [61. 19.]
 [ 9. 61.]
 [39.  2.]
 [61. 18.]
 [ 4.  9.]
 [ 5. 12.]
 [ 1. 12.]
 [ 4.  9.]
 [54.  7.]
 [61. 54.]
 [ 1. 61.]
 [14.  1.]
 [ 4. 14.]
 [61.  4.]
 [23. 61.]
 [ 9. 39.]
 [39.  8.]
 [ 8.  5.]
 [61. 61.]
 [ 9. 19.]
 [39. 12.]
 [19.  9.]
 [61. 39.]
 [ 8.  8.]
 [ 5. 25.]
 [ 1. 61.]
 [ 4. 39.]

In [12]:
top_secret = top_secret_matrix.flatten(order="F")
print(top_secret.shape)

(955,)


In [13]:
checkpoint_2 = np.all(top_secret[500:] < 100)

print(checkpoint_2)

True


In [14]:
top_secret[37:465] = top_secret[37:465] ** 0.5
top_secret = np.round(top_secret)
top_secret[top_secret == 39] = 20

In [15]:
checkpoint_3 = np.sum(top_secret)

In [16]:
number_of_evens = np.sum(top_secret % 2 == 0)

print("Number of even values:", number_of_evens)

Number of even values: 344


### Hint

Recall that basic math operations are **vectorized**.  This means you don't have to repeat the process for each item in a list; you can apply the math to the whole thing at once.

### Instructions

By altering this top secret set of numbers, you will be able to create a message.  Write your own code to complete the steps below.

1. Add 14 to every number.
2. Multiply every number by 18, then subtract 257.
3. Exponentiate every number.  (That is, do e^[number].  You may have to Google how to do this!)
4. Square every number.

**Checkpoint:** Headquarters has informed you that at this stage of decoding, there should be 352 numbers in the secret message that are below 17.  Write code to check if this is true for your data.


5. Turn your one-dimensional array of numbers into a two-dimensional array (matrix) with 5 columns. This conversion should be **column-wise**; that is, the first several elements of the array should all end up in the first **column** of your matrix.
6. Separately from your top secret numbers, create an array of all the even numbers between 1 and 382. Name it "evens". That is, "evens" should contain 2, 4, 6, 8 ..., 382.
7. Subtract the "evens" vector from the first column of your secret message matrix.
8. Subtract 100 from all numbers in the 18-24th rows of the 3rd column.
9. Multiply all numbers in the 4th and 5th column by 2.
10. Turn your matrix back into a one dimensional array. Again this should be **column-wise**, so that the first column of the matrix becomes the first elements of the array.


**Checkpoint:** Headquarters has informed you that at this stage of decoding, all numbers in indices 500 and beyond are below 100.  Write code to check if your numbers fulfill this requirement.

11. Take the square root of all numbers in indices 37 to 464.
12. Round all numbers to the nearest whole number.
13. Replace all instances of the number 39 with 20.

**Checkpoint:** Headquarters has informed you that your final message should have 344 even numbers.  Write code to check if your numbers fulfill this requirement.

## Part 3: The secret message!

Run the following to discover the final message!  Google the first line of this message, if you do not recognize it, to see what it is.


In [20]:
''.join(map(str, symbols_df.loc[top_secret-1]['Symbol']))

'\'Twas brillig, and the slithy toves   Did gyre and gimble in the wabe: All mimsy were the borogoves,   And the mome raths outgrabe. "Beware the Jabberwock, my son!   The jaws that bite, the claws that catch! Beware the Jubjub bird, and shun   The frumious Bandersnatch!" He took his vorpal sword in hand;   Long time the manxome foe he sought- So rested he by the Tumtum tree   And stood awhile in thought. And, as in uffish thought he stood,   The Jabberwock, with eyes of flame, Came whiffling through the tulgey wood,   And burbled as it came! One, two! One, two! And through and through   The vorpal blade went snicker-snack! He left it dead, and with its head   He went galumphing back. "And hast thou slain the Jabberwock?   Come to my arms, my beamish boy! O frabjous day! Callooh! Callay!"   He chortled in his joy. \'Twas brillig, and the slithy toves   Did gyre and gimble in the wabe: All mimsy were the borogoves,   And the mome raths outgrabe.'

In [ ]:
# The opening line from the poem "Jabberwocky" from the book "Through the Looking Glass"